File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/dialog_custom/CustomDialog.java
package com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ProgressBar;

import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils;
import com.github.rooneyandshadows.lightbulb.dialogs.R;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;

import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.LinearLayoutCompat;
import androidx.constraintlayout.widget.ConstraintLayout;

public class CustomDialog extends BaseDialogFragment {
    private static final String IS_LOADING_KEY = "IS_LOADING_KEY";
    private ProgressBar loadingIndicator;
    private AppCompatTextView titleView;
    private CustomDialogInflater dialogInflater;
    private boolean loading = false;

    public static CustomDialog newCustomDialogInstance(
            String title, String message, DialogButtonConfiguration positive, DialogButtonConfiguration negative,
            boolean cancelable, boolean loading, DialogTypes dialogType, DialogAnimationTypes animationType) {
        CustomDialog f = new CustomDialog();
        DialogBundleHelper helper = new DialogBundleHelper()
                .withTitle(title)
                .withMessage(message)
                .withPositiveButtonConfig(positive)
                .withNegativeButtonConfig(negative)
                .withCancelable(cancelable)
                .withShowing(false)
                .withDialogType(dialogType == null ? DialogTypes.NORMAL : dialogType)
                .withAnimation(animationType == null ? DialogAnimationTypes.NO_ANIMATION : animationType);
        helper.getBundle().putBoolean(IS_LOADING_KEY, loading);
        f.setArguments(helper.getBundle());
        return f;
    }

    @Override
    protected final View setDialogLayout(LayoutInflater inflater) {
        View view = View.inflate(getContext(), R.layout.dialog_custom, null);
        LinearLayoutCompat contentContainer = view.findViewById(R.id.customDialogContentContainer);
        contentContainer.removeAllViews();
        if (getDialogType() == DialogTypes.FULLSCREEN)
            contentContainer.getLayoutParams().height = 0;
        if (dialogInflater != null)
            contentContainer.addView(dialogInflater.inflateView(this, inflater));
        return view;
    }

    @Override
    protected void create(Bundle dialogArguments, Bundle savedInstanceState) {
        loading = savedInstanceState == null ?
                dialogArguments.getBoolean(IS_LOADING_KEY) :
                savedInstanceState.getBoolean(IS_LOADING_KEY);
    }

    @Override
    protected void configureContent(View view, Bundle savedInstanceState) {
        selectViews();
        setupLoadingView();
    }

    @Override
    protected void saveInstanceState(Bundle outState) {
        super.saveInstanceState(outState);
        outState.putBoolean(IS_LOADING_KEY, loading);
    }

    public final void setLoading(boolean isLoading) {
        this.loading = isLoading;
        setupLoadingView();
    }

    public final void setDialogInflater(CustomDialogInflater dialogInflater) {
        this.dialogInflater = dialogInflater;
    }

    public interface CustomDialogInflater {
        View inflateView(CustomDialog dialog, LayoutInflater layoutInflater);
    }

    private void selectViews() {
        if (getView() == null)
            return;
        loadingIndicator = getView().findViewById(R.id.loadingIndicator);
        titleView = getView().findViewById(R.id.title);
    }

    private void setupLoadingView() {
        if (loadingIndicator == null)
            return;
        loadingIndicator.setVisibility(loading ? View.VISIBLE : View.GONE);
        int endPadding = loading ? ResourceUtils.getDimenPxById(loadingIndicator.getContext(), R.dimen.dialog_spacing_size_small) : 0;
        titleView.setPadding(titleView.getPaddingLeft(), titleView.getPaddingTop(), endPadding, titleView.getPaddingBottom());
    }
}

File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/dialog_custom/CustomDialog.kt
package com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.ProgressBar
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.appcompat.widget.AppCompatTextView
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils
import com.github.rooneyandshadows.lightbulb.dialogs.R
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*

@Suppress("unused")
open class CustomDialog : BaseDialogFragment() {
    private var loadingIndicator: ProgressBar? = null
    private var titleView: AppCompatTextView? = null
    private var dialogInflater: CustomDialogInflater? = null
    var isLoading = false
        set(value) {
            field = value
            setupLoadingView()
        }

    companion object {
        private const val IS_LOADING_KEY = "IS_LOADING_KEY"
        fun newCustomDialogInstance(): CustomDialog {
            return CustomDialog()
        }
    }

    @Override
    override fun getDialogLayout(layoutInflater: LayoutInflater): View {
        val view = View.inflate(context, R.layout.dialog_custom, null)
        val contentContainer = view.findViewById<LinearLayoutCompat>(R.id.customDialogContentContainer)
        contentContainer.removeAllViews()
        if (dialogType == DialogTypes.FULLSCREEN) contentContainer.layoutParams.height = 0
        dialogInflater?.apply {
            val content = inflateView(this@CustomDialog, layoutInflater)
            contentContainer.addView(content)
        }
        return view
    }

    @Override
    override fun doOnCreate(dialogArguments: Bundle?, savedInstanceState: Bundle?) {
        isLoading = savedInstanceState?.getBoolean(IS_LOADING_KEY)
            ?: dialogArguments!!.getBoolean(IS_LOADING_KEY)
    }

    @Override
    override fun configureContent(view: View, savedInstanceState: Bundle?) {
        selectViews()
        setupLoadingView()
    }

    @Override
    override fun doOnSaveInstanceState(outState: Bundle?) {
        super.doOnSaveInstanceState(outState)
        outState!!.putBoolean(IS_LOADING_KEY, isLoading)
    }

    fun setDialogInflater(dialogInflater: CustomDialogInflater?) {
        this.dialogInflater = dialogInflater
    }

    interface CustomDialogInflater {
        fun inflateView(dialog: CustomDialog?, layoutInflater: LayoutInflater?): View?
    }

    private fun selectViews() {
        if (view == null) return
        loadingIndicator = requireView().findViewById(R.id.loadingIndicator)
        titleView = requireView().findViewById(R.id.dialogTitleTextView)
    }

    private fun setupLoadingView() {
        if (loadingIndicator == null) return
        loadingIndicator!!.visibility = if (isLoading) View.VISIBLE else View.GONE
        val endPadding =
            if (isLoading) ResourceUtils.getDimenPxById(loadingIndicator!!.context, R.dimen.dialog_spacing_size_small) else 0
        titleView!!.setPadding(titleView!!.paddingLeft, titleView!!.paddingTop, endPadding, titleView!!.paddingBottom)
    }
}